### ISSUE TITLE
Loading indicator is shown indefinitely and no synchronization

### ISSUE THREAD

**Issue by wiwie** [stefan-niedermann/nextcloud-deck#680]
Android deck version: 1.9.7
Server deck version: 1.0.5

We often have the problem that when editing a card and navigating back to the card overview, the loading indicator is shown indefinitely but apparently no synchronization is happening. Only when forcefully closing the app and reopening it the edits of other persons are shown.

**Comment by stefan-niedermann** [comment]
The sync is currently pretty slow because all the data have to be transferred completely. A little server hickup might interrupt this process. We are currently working on `ETag` support which will drastically speed up the sync and save much data transfer. If you agree, I propose to close this as it is nearly impossible to debug and would like to ask you to report the problem if it still occurs with the app once you read `ETag`-support in the release notes?

**Comment by wiwie** [comment]
Yes, i totally agree that it's almost impossible to debug as is. I think before closing this I'll look for something in the logs?

**Comment by stefan-niedermann** [comment]
Yeah, feel free to install an ADB-app to read the debug logs 🙂

**Comment by jakobroehrl** [comment]
I have exactly this problem when I open the deck app and do a reload. The app was running in the background.
Killing the app "solves" this issue. Its not a performance issue, I only have some cards, after killing the app the sync only takes 1-2 secs

**Comment by maxammann** [comment]
I have the same problem, but I think that there could be potentially a bug. I watched the ADB logcat while refreshing the app.

There is a lot of output for each request the app does. This stops after half a minute, but the loading spinner is still visible. The spinner keeps showing for at least 10min. The view of the board seems up to date.

This could mean that the loading spinner is not notified about the finished sync.

**Comment by stefan-niedermann** [comment]
It might be caused by the queueing mechanism. If one opens the app it will always perform a sync (1). If one does a pull to refresh, a second sync (2) is triggered, but it will wait until the (1) first sync finished and then start.

Maybe there is some kind of conflict under special circumstances which prevents the "end" event to reach the UI or so. 🤔

**Comment by maxammann** [comment]
I will take a look at the logcat logs in the following days/weeks

**Comment by desperateCoder** [comment]
@maxammann totally appreciated! It's hard to fix those kind of bugs blindly. Thanks for your help and thanks for the report @wiwie !

**Comment by maxammann** [comment]
Here is a log, which shows a problem. The server responds with a 304 in the last line. The loading spinner does not disappear and a later refresh is not possible.

```
09-12 10:57:30.826  8023  8023 D DeckLog : --- Read: shared_preference_theme_main
09-12 10:57:30.826  8023  8023 I chatty  : uid=10200(it.niedermann.nextcloud.deck) identical 2 lines
09-12 10:57:30.826  8023  8023 D DeckLog : --- Read: shared_preference_theme_main
09-12 10:57:30.838  8023  8023 E RecyclerView: No adapter attached; skipping layout
09-12 10:57:30.849  8023  8023 D DeckLog : --- Read: shared_preference_last_account | 2
09-12 10:57:30.849  8023  8023 I DeckLog : Syncing the current account on app start
09-12 10:57:30.850  8023  8023 D DeckLog : --- Write: shared_preference_last_account | 2
09-12 10:57:30.851  8023  8023 D DeckLog : --- Read: shared_preference_last_board_for_account_2 | 24
09-12 10:57:30.851  8023  8055 D DeckLog : Got Network connection
09-12 10:57:30.853  8023  8023 V DeckLog : Displaying maintenance mode info for max@cloud.maxammann.org: false
09-12 10:57:30.856  8023  8023 D DeckLog : --- Write: shared_preference_theme_main | -13134676
09-12 10:57:30.856  8023  8023 D DeckLog : --- Write: shared_preference_last_board_for_account_2 | 24
09-12 10:57:30.861  8023  8023 D DeckLog : --- Read: shared_preference_last_stack_for_account_and_board2_24 | 137
09-12 10:57:30.877  8023  8023 D DeckLog : --- Read: shared_preference_theme_main
09-12 10:57:30.881  8023  8023 D DeckLog : --- Read: shared_preference_theme_main
09-12 10:57:30.885  8023  8023 D DeckLog : --- Read: shared_preference_theme_main
09-12 10:57:30.891  8023  8023 D DeckLog : --- Write: shared_preference_last_stack_for_account_and_board_2_24 | 137
09-12 10:57:40.736  8023  8023 I DeckLog : Clearing Glide memory cache
09-12 10:57:40.738  8023 21536 I DeckLog : Clearing Glide disk cache
09-12 10:57:40.742  8023 21538 D com.nextcloud.android.sso.api.NetworkRequest: [connectApiWithBackoff] connectApiWithBackoff() called from Thread: [Thread-206]
09-12 10:57:40.743  8023 21537 D com.nextcloud.android.sso.api.NextcloudRetrofitServiceMethod: NextcloudRetrofitServiceMethod() called with: apiEndpoint = [/ocs/v2.php/], method = [public abstract io.reactivex.Observable it.niedermann.nextcloud.deck.api.NextcloudServerAPI.getCapabilities(java.lang.String)]
09-12 10:57:40.743  8023 21537 D com.nextcloud.android.sso.api.NextcloudRetrofitServiceMethod: invoke call to api using observable com.nextcloud.android.sso.api.ParsedResponse<it.niedermann.nextcloud.deck.model.ocs.Capabilities>
09-12 10:57:40.744  8023 21539 V com.nextcloud.android.sso.api.AidlNetworkRequest: [waitForApi] - api not ready yet.. waiting [RxNewThreadScheduler-1132]
09-12 10:57:40.751  8023  8023 D com.nextcloud.android.sso.api.NetworkRequest: [connectApiWithBackoff] trying to connect..
09-12 10:57:40.751  8023  8023 D com.nextcloud.android.sso.api.AidlNetworkRequest: [connect] Binding to AccountManagerService for type [nextcloud]
09-12 10:57:40.751  8023  8023 D com.nextcloud.android.sso.api.NetworkRequest: [connect] connect() called [main] Account-Type: [nextcloud]
09-12 10:57:40.751  8023  8023 D com.nextcloud.android.sso.api.AidlNetworkRequest: [connect] Component name is: [com.nextcloud.client]
09-12 10:57:40.752  8023  8023 D com.nextcloud.android.sso.api.AidlNetworkRequest: [connect] Bound to AccountManagerService successfully
09-12 10:57:40.752  8023  8023 D com.nextcloud.android.sso.api.AidlNetworkRequest: [onServiceConnected] called from Thread: [main] with IBinder [ComponentInfo{com.nextcloud.client/com.owncloud.android.services.AccountManagerService}]: android.os.BinderProxy@66f7c1
09-12 10:57:40.759  8023 21540 D com.nextcloud.android.sso.api.AidlNetworkRequest: copy data from service finished
09-12 10:57:41.063  8023 21539 V DeckLog : HTTP Status 304: There haven't been any changes on the server side for this request.
```

**Comment by creywood** [comment]
> I have exactly this problem when I open the deck app and do a reload. The app was running in the background.
> Killing the app "solves" this issue. Its not a performance issue, I only have some cards, after killing the app the sync only takes 1-2 secs

I'm also seeing this. 

If you need any further logs etc let me know.

**Comment by varac** [comment]
I also often experience this. I close the app, start it again, swip down to start a sync, and the cards don't sync. The only workaround I found is to force-stop the app from the system settings and delete the cache. Afterwards it syncs initially and then the loading indicator keeps spinning forever again. 
This make the app pretty unusable for me.
Is there any issue to track the `etag` progress ?

**Comment by varac** [comment]
Nevermind, found these related etag issues: [android app](https://github.com/stefan-niedermann/nextcloud-deck/issues/359) which relies on the [server app](https://github.com/nextcloud/deck/issues/1650). Fingers crossed and thanks anyway for this nice app !

**Comment by stefan-niedermann** [comment]
`ETag`s have been implemented and will be shipped with `1.13.0`. They require at least version [`1.2.0` of the server app](https://github.com/nextcloud/deck/releases/tag/v1.2.0). This should completely solve this issue - please feel free to create a new or reopen this issue if the problem appears again with `1.13.0` (or higher) of the Android app. (Staged Roll-Out will start soon...)

**Comment by jakobroehrl** [comment]
Working superfast and without any problems with the new 1.13!
👍

**Comment by jakobroehrl** [comment]
@stefan-niedermann 
With 1.2.3 the problem is back.
Could I provide more informations to help?

**Comment by jakobroehrl** [comment]
@stefan-niedermann 
Found the reason for the my sync problem. 
Add 2 accounts and sync, delete the deck app on one of the the two servers. Now the sync doesn't work for both added accounts.

**Comment by maxammann** [comment]
@jakobroehrl I only use one account and did not have problems the past months. Seems like the issue you are experiencing is a different one. I think you should create a separate issue.

**Comment by varac** [comment]
This is still not solved with latest v1.14.6. I had this for-ever spinning wheel multiple time in the last days.
When I force-stop the app and start again, it syncs very fast.

**Comment by szaimen** [comment]
If I understand this issue correctly is that the Deck app automatically starts the synchronisation when the app is started. If you then pull to refresh manually, it justs shows the spinner forever because there is probably some race condition between the manual and automatic refresh. How this could be solved is to show the spinner always if the synchronisation happens so that you don't trigger a manual refresh, when a refresh is already happening in the background. 
@stefan-niedermann what do you think abour this idea?

**Comment by stefan-niedermann** [comment]
> If I understand this issue correctly is that the Deck app automatically starts the synchronisation when the app is started.

This is correct.

> How this could be solved is to show the spinner always if the synchronisation happens so that you don't trigger a manual refresh, when a refresh is already happening in the background.

Actually we are queuing the synchronization attempts - so it will be filled at start with the startup-sync. If one then triggers a manual sync, it will be queued and executed after the startup-sync has been finished. I agree there might be some race condition - the question is whether it is only a visual issue or a sync issue.

@desperateCoder i guess the cleanest way to fix any visual glitches is to expose a `LiveData<boolean> isSynchronizationActive` which i subscribe to in the UI instead of handling the state manually. What do you think?

**Comment by szaimen** [comment]
Could you maybe reopen this issue until an answer from @desperateCoder is received?

**Comment by stefan-niedermann** [comment]
Okay, i finally found the one scenario when the loading indicator is shown indefinitely. This can be caused by a not caught exception when refreshing the capabilities. In case of an error, it was silently logged and the spinner was not stopped.

I fixed this by hiding the spinner properly now *and* display an exception dialog with the information about what went actually wrong.

I closed this issue because we now can tackle the actual problems (which might be different for each and every one of the people who reported the issue) and the actual issue ("Loading indicator is shown indefinitely") is solved.
Please feel free to report the problems you are facing in separate new issues as soon as you can see them.

This fix will be in version `1.15.6` and soon released on F-Droid (Play Store will take more time). I will also post here a link as soon as the build server has finished the build, so whoever doesn't want to wait but wants to test the new version will be able to install it in parallel to the currently released versions.

**Comment by stefan-niedermann** [comment]
Dev build from the build server: https://github.com/stefan-niedermann/nextcloud-deck/suites/2305177928/artifacts/48363798

**Comment by birefringence** [comment]
I've tried the dev build, but I don't see any difference. The spinner stays, there is no sync, and I don't get any error message.
So far, this has only happend after I changed some cards on the server.
Can I provide any logs? (I should have a working adb)

**Comment by desperateCoder** [comment]
> Can I provide any logs? 

Woah, that would be awesome! I'd check if there is enough logging beforehand though, I'll let you know!

**Comment by stefan-niedermann** [comment]
@birefringence [Here is a build with excessive logging](https://github.com/stefan-niedermann/nextcloud-deck/suites/2308223002/artifacts/48442370). Would you please install it?
Just wait 10-20 seconds after the spinner appears and then press the bug button in the bottom right corner. The logs should be copied directly into your clipboard. Please make sure to remove any sensitive data and then send it to info@niedermann.it or even better post it here (if you can).

**Comment by birefringence** [comment]
Thanks! I've installed the build with logging. However, I haven't been able to trigger this problem anymore since then, although I've been editing and syncing cards all evening. I'll keep using this build for now and see if hit it again.

**Comment by stefan-niedermann** [comment]
@creywood @varac @maxammann and every other who upvoted the issue: maybe one of you can verify whether or not the spinner issue was solved [with this build](https://github.com/stefan-niedermann/nextcloud-deck/issues/680#issuecomment-803430532) and gather us some logs if not [with this build](https://github.com/stefan-niedermann/nextcloud-deck/issues/680#issuecomment-803562861)?

> I've tried the dev build, but I don't see any difference

@birefringence i am wondering if you, by any chance, didn't launch / use the dev build but the released app maybe?

**Comment by birefringence** [comment]
Ah well, I think I experienced again. But I couldn't get the log, because this happened when I hit the button:
```
Full Crash:

App Version: 1.15.6
App Version Code: 1015006
App Flavor: dev

Files App Version Code: 30150190

---

OS Version: 4.4.205-gf0aab43(cabd69955f)
OS API Level: 29
Device: H3113
Manufacturer: Sony
Model (and Product): H3113 (H3113)

---

java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 9016864 bytes
	at android.content.ClipboardManager.setPrimaryClip(ClipboardManager.java:108)
	at it.niedermann.android.util.ClipboardUtil.copyToClipboard(ClipboardUtil.kt:26)
	at it.niedermann.android.util.ClipboardUtil.copyToClipboard(ClipboardUtil.kt:15)
	at it.niedermann.nextcloud.deck.ui.MainActivity.lambda$onCreate$8$MainActivity(MainActivity.java:329)
	at it.niedermann.nextcloud.deck.ui.-$$Lambda$MainActivity$TS-zapqfS60RYdR-AO2LkTGJi0s.onClick(Unknown Source:2)
	at android.view.View.performClick(View.java:7259)
	at android.view.View.performClickInternal(View.java:7236)
	at android.view.View.access$3600(View.java:801)
	at android.view.View$PerformClick.run(View.java:27892)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7356)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)
Caused by: android.os.TransactionTooLargeException: data parcel size 9016864 bytes
	at android.os.BinderProxy.transactNative(Native Method)
	at android.os.BinderProxy.transact(BinderProxy.java:511)
	at android.content.IClipboard$Stub$Proxy.setPrimaryClip(IClipboard.java:288)
	at android.content.ClipboardManager.setPrimaryClip(ClipboardManager.java:106)
	... 15 more

**Comment by stefan-niedermann** [comment]
Shit! The log was too long and could not be copied to the clipboard. Let me think about what else we can do to get those logs.

**Comment by stefan-niedermann** [comment]
So i made [another build](https://github.com/stefan-niedermann/nextcloud-deck/suites/2314333790/artifacts/48607677) which has a few enhancements:
1. The log will no longer be copied to the clipboard but shared **as a file** (→ no length limitation). You should be able to share the log file with any other app (I tested E-Mail, Telegram and Nextcloud)
2. To activate the logging, go to the settings and enable "Debug logs"
3. To copy the collected logs, use the bug icon in the sidebar header (only visible after step 2)
  ![grafik](https://user-images.githubusercontent.com/4741199/111985069-7690c680-8b0c-11eb-9f14-41fa50d6502f.png)

**Comment by birefringence** [comment]
Ok, got one. I hope I didn't redact it too much. As far as I can tell from the server logs, the app didn't reach the server when trying manually refresh by pulling.
I sent an e-mail since I can't attach files here (or at least I don't know how).

**Comment by stefan-niedermann** [comment]
Interesting. `SyncManager#refreshCapabilities()` does never call the `onResponse` or the `onError` callback.

@desperateCoder i think we are missing a generic `else` case here for the `NextcloudHttpRequestFailedException` in https://github.com/stefan-niedermann/nextcloud-deck/blob/845cbaf31e357af7f3d80b3bfb2676fa2ccfb935/app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/SyncManager.java#L451

What do you think?

**Comment by desperateCoder** [comment]
Id suspect a missing `callback.` before this line:

https://github.com/stefan-niedermann/nextcloud-deck/blob/845cbaf31e357af7f3d80b3bfb2676fa2ccfb935/app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/SyncManager.java#L458

Why doesn't that end in an infinite loop tho? I'll have a look at lunch break... Thanks for the research!

**Comment by desperateCoder** [comment]
@birefringence we had two blind spots here, we just fixed them. 

Please try this one: [(｡◕‿‿◕｡)](https://github.com/stefan-niedermann/nextcloud-deck/suites/2323602703/artifacts/48890710)

This time ther shoud (TM) come up an error, in case something goes wrong at sync. Maybe thats already the solution to the problem, but yeah, lets see...

Thanks for the research and testing btw, this helps us a LOT!

**Comment by birefringence** [comment]
Well, the spinner now does disappear, at least after I switch away from the app and come back. But there is still no sync. Here is the log:
```
2021-03-23 18:17:27 INFO (SyncWorker.java:39) doWork() → Starting background synchronization
2021-03-23 18:17:31 VERBOSE (SyncManager.java:451) onError() → HTTP Status 304
2021-03-23 18:17:31 VERBOSE (SyncManager.java:464) onError() → → There haven't been any changes on the server side for this request.
2021-03-23 18:17:31 DEBUG (SyncHelper.java:86) onError() → [BoardDataProvider] ETags do match! skipping this one.
2021-03-23 18:17:31 INFO (SyncManager.java:287) respondCallbacksAfterSync() → SyncQueue: responding sync for 1 queued callbacks!
2021-03-23 18:17:31 INFO (SyncWorker.java:43) doWork() → Finishing background synchronization with result false
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:42 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:43 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:32:50 INFO (SyncWorker.java:39) doWork() → Starting background synchronization
2021-03-23 18:42:15 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:42:15 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:42:15 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:42:15 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:42:15 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:42:16 DEBUG (BrandingUtil.java:56) saveBrandColors() → --- Write: shared_preference_theme_main | -16742459
2021-03-23 18:42:16 DEBUG (DeckApplication.java:106) saveCurrentBoardId() → --- Write: shared_preference_last_board_for_account_1 | 1
2021-03-23 18:42:16 DEBUG (DeckApplication.java:128) readCurrentStackId() → --- Read: shared_preference_last_stack_for_account_and_board1_1 | 20
2021-03-23 18:42:16 INFO (SyncWorker.java:39) doWork() → Starting background synchronization
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:18 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:42:49 DEBUG (MainActivity.java:914) onLost() → Network lost
2021-03-23 18:43:46 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:43:46 DEBUG (DeckApplication.java:100) readCurrentAccountId() → --- Read: shared_preference_last_account | 1
2021-03-23 18:43:46 INFO (MainActivity.java:224) lambda$onCreate$15$MainActivity() → Syncing the current account on app start
2021-03-23 18:43:46 DEBUG (MainActivity.java:897) onAvailable() → Got Network connection
2021-03-23 18:43:46 DEBUG (DeckApplication.java:92) saveCurrentAccountId() → --- Write: shared_preference_last_account | 1
2021-03-23 18:43:46 DEBUG (DeckApplication.java:114) readCurrentBoardId() → --- Read: shared_preference_last_board_for_account_1 | 1
2021-03-23 18:43:46 VERBOSE (MainActivity.java:300) lambda$onCreate$7$MainActivity() → Displaying maintenance mode info for user@domain.de/nextcloud: false
2021-03-23 18:43:47 DEBUG (BrandingUtil.java:56) saveBrandColors() → --- Write: shared_preference_theme_main | -16742459
2021-03-23 18:43:47 DEBUG (DeckApplication.java:106) saveCurrentBoardId() → --- Write: shared_preference_last_board_for_account_1 | 1
2021-03-23 18:43:47 DEBUG (DeckApplication.java:128) readCurrentStackId() → --- Read: shared_preference_last_stack_for_account_and_board1_1 | 20
2021-03-23 18:43:47 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:43:47 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:43:47 DEBUG (BrandingUtil.java:43) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-03-23 18:43:47 DEBUG (DeckApplication.java:120) saveCurrentStackId() → --- Write: shared_preference_last_stack_for_account_and_board_1_1 | 20
2021-03-23 18:43:57 INFO (MainActivity.java:380) lambda$onCreate$14$MainActivity() → > Triggered manual refresh
2021-03-23 18:43:57 INFO (MainActivity.java:385) lambda$onCreate$14$MainActivity() → Clearing Glide memory cache
2021-03-23 18:43:57 INFO (MainActivity.java:395) lambda$onCreate$14$MainActivity() → > refreshCapabilities()
2021-03-23 18:43:57 INFO (MainActivity.java:388) lambda$onCreate$13$MainActivity() → Clearing Glide disk cache
2021-03-23 18:43:57 INFO (MainActivity.java:512) refreshCapabilities() → > refreshCapabilities()
2021-03-23 18:43:57 INFO (MainActivity.java:397) lambda$onCreate$14$MainActivity() → > mainViewModel.synchronize()
2021-03-23 18:43:57 VERBOSE (SyncManager.java:451) onError() → HTTP Status 304
2021-03-23 18:43:57 VERBOSE (SyncManager.java:464) onError() → → There haven't been any changes on the server side for this request.
2021-03-23 18:43:57 INFO (MainActivity.java:516) onResponse() → > refreshCapabilities() → onResponse
2021-03-23 18:43:57 INFO (MainActivity.java:521) onResponse() → > refreshCapabilities() → onResponse → maintenance enabled: true

**Comment by stefan-niedermann** [comment]
I think we're coming closer. [Give it another try](https://github.com/stefan-niedermann/nextcloud-deck/suites/2333166147/artifacts/49177248), please (And please **don't disable** the new `ETag` setting :smile:)

> But there is still no sync.

Can you elaborate please a bit more what "there is still no sync" does mean for you? The sync doesn't work at all? It does only work in one direction (e. g. Android -> Cloud)? It does only not work in specific cases (like adding an attachment or something else via the Web UI)? Have you seen any error message (dialog or a snackbar at the bottom)?

**Comment by birefringence** [comment]
Thanks I'll try again! So far I mostly changed things on the server, which then didn't shop up in the app until I forcefully closed and restarted it. I will specifically try to edit cards in the app and see if it goes through the other way. Most of the changes were creating and moving cards, and so far I haven't noticed that this phenomenon only applied to specific types of changes. There were no error messages.

**Comment by birefringence** [comment]
Oh no, I just noticed I made a mistake when copying the APK. I think I was still running the previous version, when I created the last log. Sorry! I'm now definitely on the latest version and I will continue to test.

**Comment by desperateCoder** [comment]
@birefringence I'd bet at least a used tissue, that the error won't happen anymore. @stefan-niedermann isn't convinced yet, but if you can't reproduce the issue anymore, maybe my changes kicked in. Looking forward to your feedback! :blush: 

> Sorry!

No need to be sorry here! You already helped us a lot! Thanks for that! :heart:

**Comment by stefan-niedermann** [comment]
Most of the improvements which we made here have made it into version `1.15.8` (released on F-Droid and Play Store Beta channel, should arrive in the next few days). I would appreciate any feedback with this released version.

I kept one change back: @desperateCoder suggested to remove our queued synchronization (which holds back triggered synchronizations if there is still a sync in progress). I'd like to see whether or not the endless spinner still occurs and if yes, i will also merge this change. Though i'd like to keep this feature, so let's wait for some feedback of `1.15.8`. :slightly_smiling_face:

**Comment by birefringence** [comment]
I haven't had any problems with the last dev build from this thread. I will also try 1.15.8 as soon as possible.

**Comment by birefringence** [comment]
I'm currently on the 1.16.0 from the Google Play beta channel. Today the spinner stuck again. Unfortunately, I didn't have the log collection switch on until after it happened. So the log might be missing relevant information:
```
2021-04-03 14:28:42 DEBUG (SettingsFragment.java:82) lambda$onCreatePreferences$4() → persistet debug logs: true
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:43 INFO (FilterViewModel.java:122) setFilterText() → New filterText: 
2021-04-03 14:28:43 DEBUG (DeckApplication.java:122) readCurrentAccountId() → --- Read: it.niedermann.nextcloud.deck.last_account → 1
2021-04-03 14:28:43 INFO (MainActivity.java:248) lambda$onCreate$16$MainActivity() → Syncing the current account on app start
2021-04-03 14:28:43 DEBUG (MainActivity.java:951) onAvailable() → Got Network connection
2021-04-03 14:28:43 DEBUG (DeckApplication.java:100) saveCurrentAccount() → --- Write: it.niedermann.nextcloud.deck.last_account → 1
2021-04-03 14:28:43 DEBUG (DeckApplication.java:102) saveCurrentAccount() → --- Write: it.niedermann.nextcloud.deck.last_account_color → -16743735
2021-04-03 14:28:43 DEBUG (DeckApplication.java:136) readCurrentBoardId() → --- Read: it.niedermann.nextcloud.deck.last_board_for_account_1 → 1
2021-04-03 14:28:43 VERBOSE (MainActivity.java:324) lambda$null$8$MainActivity() → Displaying maintenance mode info for user@domain.com/nextcloud: false
2021-04-03 14:28:43 DEBUG (BrandingUtil.java:47) saveBrandColors() → --- Write: shared_preference_theme_main | -16742459
2021-04-03 14:28:43 DEBUG (DeckApplication.java:128) saveCurrentBoardId() → --- Write: it.niedermann.nextcloud.deck.last_board_for_account_1 → 1
2021-04-03 14:28:43 DEBUG (DeckApplication.java:150) readCurrentStackId() → --- Read: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 20
2021-04-03 14:28:44 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:44 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:44 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:28:44 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 20
2021-04-03 14:29:07 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:07 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 1
2021-04-03 14:29:08 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 20
2021-04-03 14:29:09 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 1
2021-04-03 14:29:10 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 2
2021-04-03 14:29:10 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 3
2021-04-03 14:29:11 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 2
2021-04-03 14:29:12 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 1
2021-04-03 14:29:12 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 20
2021-04-03 14:29:13 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:38 DEBUG (MainActivity.java:951) onAvailable() → Got Network connection
2021-04-03 14:29:38 DEBUG (MainActivity.java:951) onAvailable() → Got Network connection
2021-04-03 14:29:38 DEBUG (MainActivity.java:951) onAvailable() → Got Network connection
2021-04-03 14:29:51 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:51 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:51 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:51 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:51 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:52 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 1
2021-04-03 14:29:53 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 2
2021-04-03 14:29:53 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 3
2021-04-03 14:29:54 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 2
2021-04-03 14:29:54 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 1
2021-04-03 14:29:54 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 20
2021-04-03 14:29:58 DEBUG (BrandingUtil.java:47) saveBrandColors() → --- Write: shared_preference_theme_main | -6355573
2021-04-03 14:29:58 DEBUG (DeckApplication.java:128) saveCurrentBoardId() → --- Write: it.niedermann.nextcloud.deck.last_board_for_account_1 → 5
2021-04-03 14:29:58 DEBUG (DeckApplication.java:150) readCurrentStackId() → --- Read: it.niedermann.nextcloud.deck.last_stack_for_board_1_5 → 21
2021-04-03 14:29:58 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:58 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:29:58 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:01 DEBUG (BrandingUtil.java:47) saveBrandColors() → --- Write: shared_preference_theme_main | -16742459
2021-04-03 14:30:01 DEBUG (DeckApplication.java:128) saveCurrentBoardId() → --- Write: it.niedermann.nextcloud.deck.last_board_for_account_1 → 1
2021-04-03 14:30:01 DEBUG (DeckApplication.java:150) readCurrentStackId() → --- Read: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 20
2021-04-03 14:30:01 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:01 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:01 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:14 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:14 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:14 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:14 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:14 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:17 DEBUG (MainActivity.java:968) onLost() → Network lost
2021-04-03 14:30:17 DEBUG (MainActivity.java:968) onLost() → Network lost
2021-04-03 14:30:17 DEBUG (MainActivity.java:968) onLost() → Network lost
2021-04-03 14:30:32 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:32 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:32 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:32 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:32 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:30:44 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 1
2021-04-03 14:30:45 DEBUG (DeckApplication.java:142) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 20
2021-04-03 14:31:48 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:31:48 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:31:48 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:31:48 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main
2021-04-03 14:31:48 DEBUG (BrandingUtil.java:37) readBrandMainColor() → --- Read: shared_preference_theme_main

**Comment by stefan-niedermann** [comment]
@desperateCoder are we missing an `onError` callback for [this `onResponse`](https://github.com/stefan-niedermann/nextcloud-deck/blob/master/app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/SyncManager.java#L217)? Might also explain a not finishing spinner...

**Comment by desperateCoder** [comment]
@stefan-niedermann we do. On the other hand: The sync-queue is still there, so it doesn't surprise me, that there are still issues... We could try adding it. Or, as I'd suggest, just get rid of this whole damn thing (yep, guess I messed it up) and be happy as @birefringence can confirm. It seemed to work well, as we removed it. Am I missing something? The play store beta c

**Comment by birefringence** [comment]
1.17.7 from Google beta:
```
2021-04-18 23:16:06 DEBUG (DeckApplication.java:151) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_5 → 14
2021-04-18 23:16:07 DEBUG (DeckApplication.java:151) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_5 → 13
2021-04-18 23:16:07 DEBUG (DeckApplication.java:151) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_5 → 21
2021-04-18 23:16:20 DEBUG (BrandingUtil.java:44) saveBrandColors() → --- Write: it.niedermann.nextcloud.deck.theme_main | -16742459
2021-04-18 23:16:20 DEBUG (DeckApplication.java:137) saveCurrentBoardId() → --- Write: it.niedermann.nextcloud.deck.last_board_for_account_1 → 1
2021-04-18 23:16:20 DEBUG (DeckApplication.java:159) readCurrentStackId() → --- Read: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 1
2021-04-18 23:16:20 DEBUG (DeckApplication.java:151) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 1
2021-04-18 23:16:23 DEBUG (DeckApplication.java:151) saveCurrentStackId() → --- Write: it.niedermann.nextcloud.deck.last_stack_for_board_1_1 → 20
2021-04-18 23:16:28 INFO (MainActivity.java:405) lambda$null$14$MainActivity() → Triggered manual refresh
2021-04-18 23:16:28 INFO (CustomAppGlideModule.java:30) clearCache() → Clearing Glide memory cache
2021-04-18 23:16:28 VERBOSE (MainActivity.java:409) lambda$null$14$MainActivity() → Trigger refresh capabilities for user@domain.de/nextcloud
2021-04-18 23:16:28 INFO (CustomAppGlideModule.java:33) lambda$clearCache$0() → Clearing Glide disk cache
2021-04-18 23:16:28 VERBOSE (MainActivity.java:542) refreshCapabilities() → Refreshing capabilities for user@domain.de/nextcloud
2021-04-18 23:16:28 VERBOSE (MainActivity.java:412) lambda$null$14$MainActivity() → Trigger synchronization for user@domain.de/nextcloud
2021-04-18 23:16:28 VERBOSE (SyncManager.java:458) onError() → HTTP Status 304
2021-04-18 23:16:28 VERBOSE (SyncManager.java:470) onError() → HTTP Status 304: There haven't been any changes on the server side for this request.
2021-04-18 23:16:28 VERBOSE (MainActivity.java:546) onResponse() → Finished refreshing capabilities for user@domain.de/nextcloud successfully.
2021-04-18 23:16:28 VERBOSE (MainActivity.java:551) onResponse() → Maintenance mode is disabled → Stop spinner.
```
It says "Stop spinner" in the last line, but this is not what happened.

So far I am most likely to hit this problem, when I do a lot of changes with the web app. I then at some point notice that the cards shown in the Android app are out of sync. When I attempt to manually refresh to fix this, I only get the never-ending spinner, but the cards are not updated until I manually close the application.

**Comment by stefan-niedermann** [comment]
@desperateCoder it looks like the `SyncManager` gets a `304` while syncing the capabilities in line `470` (compare log) but the `callback.onResponse()` of line `483` never gets called...? Do you have any explanation for that? :thinking: 

https://github.com/stefan-niedermann/nextcloud-deck/blob/1.17.7/app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/SyncManager.java#L470-L484

**Comment by birefringence** [comment]
For some reason, I don't see a corresponding access with a 304 code in the server log. In fact, don't see any access at all with that time stamp.

**Comment by stefan-niedermann** [comment]
> For some reason, I don't see a corresponding access with a 304 code in the server log. In fact, don't see any access at all with that time stamp.

The requests might have hit the `capabilities` endpoint which is not necessarily in the logs (it's outside of the Deck app).

I just merged the PR of @desperateCoder which have been [confirmed to work](https://github.com/stefan-niedermann/nextcloud-deck/issues/680#issuecomment-808956700) by @birefringence and [did some other enhancements](https://github.com/stefan-niedermann/nextcloud-deck/commit/0831e0100534ebe4777ed187d354167a77a4f1ca). Let's try this in `1.18.0` :slightly_smiling_face: .

**Comment by birefringence** [comment]
This issue has not occurred again since 1.18.0. So I would consider it fixed.

**Comment by stefan-niedermann** [comment]
Thanks for the feedback 🙂 Closing this for now. Please create a new issue in case this behavior returns (this one is a bit crowded)

**Comment by desperateCoder** [comment]
So, no [used tissue](https://github.com/stefan-niedermann/nextcloud-deck/issues/680#issuecomment-805968478) for you guys, deal wit it! 

Special thanks to @birefringence for your help and responses to help us to resolve this issue!
